sdks/java/core/src/test/java/org/apache/beam/sdk/runners/TransformHierarchyTest.java

```diff
@@ -492,4 +495,87 @@ public void visitPrimitiveTransform(Node node) {
     assertThat(visitedPrimitiveNodes, containsInAnyOrder(upstreamNode, replacementParNode));
     assertThat(visitedValues, Matchers.<PValue>containsInAnyOrder(upstream, output));
   }
+
+  @Test
+  public void visitIsTopologicallyOrdered() {
```

Comment:
I've added some assertions to ensure that nodes are not visited more than once, and a test to demonstrate the children of skipped nodes are not visited.